### PR TITLE
WIP: Convert built-in readers to TypeScript

### PR DESCRIPTION
**Pull request by ericblade**
Converted barcode_reader to TS, and it seems that doing so also requires a conversion of all the subclasses, I'm pushing it here so others can jump in and help if they want.  I probably should've done this from the top level, converting the subclasses first, but oh well.

This won't pass tests until all readers are converted.

### CODE REVIEW

File: src/reader/code_93_reader.ts [hunk 1]
```diff
@@ -0,0 +1,238 @@
...
+        if (!this._verifyEnd(lastStart, nextStart)) {
+            return null;
+        }
+
+        if (!this._verifyChecksums(result)) {
+            return null;
+        }
+
+        result = result.slice(0, result.length - 2);
+        // TODO: Eric really hates assigning inside an if ugliness, but it's pretty elegant here.
```
Comment (kevinpapst) [inline src/reader/code_93_reader.ts]: Eric should either stick with his gut feelings or remove the TODO 😀

File: src/reader/2of5_reader.ts [hunk 2]
```diff
@@ -0,0 +1,224 @@
...
+    [N, N, W, N, W],
+    [W, N, W, N, N],
+    [N, W, W, N, N],
+    [N, N, N, W, W],
+    [W, N, N, W, N],
+    [N, W, N, W, N],
+];
+const START_PATTERN_LENGTH = START_PATTERN.reduce((sum, val) => sum + val, 0);
+
+class NewTwoOfFiveReader extends BarcodeReader {
```
Comment (kevinpapst) [inline src/reader/2of5_reader.ts]: Why the `New` in the class name `NewTwoOfFiveReader `? Same for the `NewCodabarReader`...

File: src/reader/code_93_reader.ts (hunk 1 above)
Comment (ericblade) [reply src/reader/code_93_reader.ts]: agree, i intend to do a full pass on any TODOs that I leave once they are completed..

File: src/reader/2of5_reader.ts (hunk 2 above)
Comment (ericblade) [reply src/reader/2of5_reader.ts]: yeah, i'll fix that on a future pass .. i'm going through each one, implementing them as a class, copying each function into the class .. named them NewWhatever to start, so my editor doesn't fill with red squiggles .. thanks :-)

**Comment by ericblade** [comment]
If anyone wants to have a look at https://github.com/ericblade/quagga2/pull/141/commits/2b4c6a635e05c66d946264f465beb6a905a3ebde#diff-6e0296ee2b29f5d8fe6163490314859d 

ean_reader is broken, but i'm not seeing an obvious point at which it differs from the original.  

i'm sure i'll figure it out sometime but if anyone else has eyes on it, let me know :-)

**Comment by ericblade** [comment]
Finally get to close this one out, yay. Now let's see if I can remember what else i was wanting to do.

**Pull request merged**